### DISCUSSION TITLE
Xamarin

### DISCUSSION THREAD

**Discussion by ghuotvez** [NRCan/GSC-Field-Application#108] [Ideas]
We'll need to look into Xamarin in order to expand the possibility of field app to be installed on other devices.

With the sudden deprecation of Panasonic thoughpads, high prices of other devices, and the talk about W11 not being available for devices with less then 9'' of screen, our field of possibilities (device wise) is getting more and more narrow.

**Comment by ghuotvez** [comment]
From a talk during today's meeting. 
@KazShimamura-GSC @rcocking @eygirard @jameljoseph

**Comment by ghuotvez** [comment]
This original idea was brought up by @KazShimamura-GSC years ago and gets more and more important as the time goes by.

We're getting new tablets in this year, but we're expecting to need another working solution in the next 3-4 years when most of our tablets gets too old.

**Reply by KazShimamura-GSC** [reply]
Or look at Flutter by Google to write cross-platform Android and iOS mobile apps, or look into low/no code applications? But Xamarin is great for C# developers!

**Reply by ghuotvez** [reply]
Thanks for the advice Kaz! I'll dig a bit into Flutter.

**Reply by ghuotvez** [reply]
Although, by using Xamarin we'll be able to keep the app on Windows and reach the other ones. Flutter seems to be made for Android and iOs only.

**Comment by ghuotvez** [comment]
https://devblogs.microsoft.com/xamarin/whats-new-in-xamarin-and-visual-studio-2022/ More neat details about Xamarin and latest VS

**Reply by ghuotvez** [reply]
Which points to Xamarin.Forms https://docs.microsoft.com/en-us/xamarin/xamarin-forms/

**Comment by ghuotvez** [comment]
This might go forward a bit more quickly then anticipated, since we might run into hardware supply for summer 2023 field trips. Not sure we could flip over the whole app before next year but it only tells us that we need to plan ahead.

**Comment by ghuotvez** [comment]
Good news being that it can support calls to the gps and maps: https://docs.microsoft.com/en-us/xamarin/essentials/ and even more, on all the platforms.

**Comment by ghuotvez** [comment]
The gift that keeps on giving, base database support for Xamarin.Forms is sqlite.

**Comment by rcocking** [comment]
Wow, you are finding so much gold at that site. It might be worth writing up a plan/proposal.

**Reply by ghuotvez** [reply]
Yeah either a plan proposal or I book the rest of my year to mostly work on this flip. In any case we need a good plan and target what we want to see first. And also check what map engine we could use. I think ESRI has a Xamarin sdk, but I'm not sure.

**Reply by KazShimamura-GSC** [reply]
If anything going forward with this start off with MVP (minimum viable product) for the app, release often, add features with each iteration - don't expect perfections but good enough is good enough ;) As long as they don't lose their data and that should be the core of MVP :) Just like the GSC Field App was focused on using SQlite instead of Shapefiles, and interface focused on easy data entry for our geologists with a textual and spatial view of the data.

**Reply by ghuotvez** [reply]
Thanks, yeah we'll need to focus on that principle.

**Comment by rcocking** [comment]
It would require a full time Gabriel. 😊 Is that even possible?

**Reply by ghuotvez** [reply]
Haha I can only wish

**Comment by eygirard** [comment]
book the rest of the year ;-)

**Comment by jameljoseph** [comment]
When I am done with the april ripley data, I will look at this Gab.

**Reply by KazShimamura-GSC** [reply]
Cool Jamel, I see that Pluralsight has many courses on Xamarin: https://app.pluralsight.com/search/?q=xamarin&m_sort=relevance&query_id=50961fcf-8446-43ee-aded-69f6ee9fe7a5&source=user_typed

**Comment by jameljoseph** [comment]
that will be my first stop when I start this

**Comment by ghuotvez** [comment]
Well, we might want to skip a step a go straight to .net MAUI, which is a better version of Xamarin, even more portable and easier to work with, meaning it will flawlessly work on any platform by default. Apparently it was a small problem with Xamarin. AT some point, MS states that MAUI will entirely replace Xamarin which will go to a deprecated state a year after last release.

> Xamarin.Forms will ship a [new major version later this year](https://aka.ms/xf-roadmap), and continue to ship minor and service releases every 6 weeks through .NET 6 GA in November 2021. The final release of Xamarin.Forms will be serviced for a year after shipping, and all modern work will shift to .NET MAUI.


More reads: 
https://ballardchalmers.com/2021/10/04/what-is-net-maui-and-how-does-it-differ-from-xamarin/
https://devblogs.microsoft.com/dotnet/introducing-net-multi-platform-app-ui/
https://tech-lead.eu/architecture/maui-vs-xamarin-forms/

**Reply by jameljoseph** [reply]
I've been doing some research on xamarin and it looks like we would have to use the forms and we might not have the same functionality or look as our app stands now so I will check out MAUI (maybe we should go to MAUI, might give us a better feel for this ui platform???)

**Reply by ghuotvez** [reply]
Awesome thanks!

**Reply by KazShimamura-GSC** [reply]
Before going deep-diving further into Maui, perhaps looking at **low code application platform** of [Microsoft PowerApps](https://powerapps.microsoft.com/en-ca/build-powerapps/) as another alternative that could reduce coding time, as well as implementing a solution that has cloud solutions baked-in or easily extend with cloud services/tools? 

I noticed there's support for [mobile canvas app](https://docs.microsoft.com/en-us/power-apps/mobile/run-powerapps-on-mobile). Maybe through PowerApps that it would make deployment and ensuring versions on Windows, Android or iOS devices are all the same for all users easier as well with Microsoft Intune? Deployment and setup take lots of effort and time as well. 

Still not sure of offline map tile support, GPS coordinate data entry, and support SQLite or what relational database is supported since I just had a high-level look at this technology?

**Reply by ghuotvez** [reply]
Good arguments, thanks Kaz for the input!

**Reply by KazShimamura-GSC** [reply]
Perhaps this is where we can establish FODR on an [Azure SQL Database](https://docs.microsoft.com/en-us/power-apps/maker/canvas-apps/app-from-azure-sql-database) right from the start so the whole data flow is centrally managed right from the get go? 

There's some [geospatial controls](https://docs.microsoft.com/en-ca/power-apps/maker/canvas-apps/geospatial-overview), too. 

I've been not paying attention to Microsoft latests technology with my management hat on most of the time, but did a quick search at [Microsoft Build Event last May](https://docs.microsoft.com/en-us/events/build-2022/?terms=power%20app) with recorded videos on Power Apps if you're interested.

**Reply by ghuotvez** [reply]
Oh yeah sure send them, I'll take a look. As for FODR, they're really aiming for a dead-end archive kind of thing rather then a start. But yeah centralizing everything should be a goal at some point. At least have some sort of conveyor to field app data up to FODR from a service or something.

**Reply by KazShimamura-GSC** [reply]
Sorry, the Microsoft Build Event videos are in my reply above as a hyper link, "Microsoft Build Event las May". It's awesome that GitHub supports markdown in these replies :)

**Reply by ghuotvez** [reply]
Yeah it is :) thanks again!

**Reply by jameljoseph** [reply]
I checked out powerapps, it seems like a very very simple yet very powerful product.  I believe Maui has been released with teh latest version of VS so we can play with both and see what their capabilities are.

**Reply by jameljoseph** [reply]
I tried signing in (not buying but just signing in) and got this:  Your organization doesn't currently allow its users to purchase Microsoft PowerApps. Contact your IT admin for more information.

**Reply by jameljoseph** [reply]
I have played with MAUI and when I try to run an android emulator I get asked for admin passwords so that is a bit of a pain

**Reply by ghuotvez** [reply]
True, I've notice that also. It is quite a pain.

**Comment by ghuotvez** [comment]
Work as started on version 3 under MAUI.

https://github.com/NRCan/GSC-Field-Application-3

**Reply by KazShimamura-GSC** [reply]
That's great news! So developing for iOS app as well for the iPhone/iPad? If so then you'll need a Mac to compile with Xcode. Lots of geologists in Vancouver prefer iOS :)

**Reply by ghuotvez** [reply]
Yeah, at some point we could even build straight to iOS given I have a mac and a dev apple license. That'll come a bit later though. For what I've been doing, trying to get everything look fine on 2 OS is something, I wouldn't want to test out a third one at the same time for now haha.
